### ISSUE TITLE
is it possible to explicit paths to bibtex and the pdfviewers ?

### ISSUE THREAD

**Issue by teto** [lervag/vimtex#1058]
# Explain the issue
I would like to explicitly set the paths to the different executables I use, i.e., set the path to the mupdf/zathura binary and the path to bibtex. I was already able to set the path to latexmk via  ` let g:vimtex_compiler_latexmk.executable`.
Is it possible to do this ? `vimtex_viewer_mupdf.executable`

For a bit of context, I use www.nixos.org, a distribution that kindof virtualizes applications by making things as explicitly as possible to prevent collisions with the environment (so a good practice is to only have the essential in PATH). As a result, reproducibility across machines is really good. This means I end up with paths like
/nix/store/df4m38i6bcpvj5kbds1rwa1dfgdfa6p4-texlive-combined-basic-2017/bin/latexmk
( ` let g:vimtex_compiler_latexmk.executable='${pkgs.texlive.combined.scheme-medium}/bin/latexmk'`)

**Comment by lervag** [comment]
Sorry for taking so long to respond!

For bibtex and similar, you can set the paths in a `.latexmkrc` file.

For Zathura and/or MuPDF, I am tempted to ask you to write your own viewer extension. It should be pretty straightforward, i.e.:

1. Copy the file `/path/to/vimtex/autoload/vimtex/view/zathura.vim` to something like `~/.vim/autoload/vimtex/view/custom.vim`.

2. Change all function names from `vimtex#view#zathura#...` to `...#custom#...`.

3. Change the desired paths accordingly.

4. Set `g:vimtex_view_method = 'custom'`.

This way you get full control of all the parts and can fix the paths that are missing.

----

So, why won't I just add the options? Well, I never implemented vimtex to support this level of customization, and I don't want to add many more options. If I were to change things now, I would need to make it quite general and add more documentation and options. I think the number of people who needs this feature (i.e. you) is very small compared to the total number of users (> 1000).

----

An alternative idea: Can't you somehow change your `$PATH` variable? Perhaps only for Vim with LaTeX files? I.e., you could do something like:

```vim
augroup vimrc
  autocmd!
  autocmd FileType tex call UpdatePath()
augroup END

function! UpdatePath() abort
  if get(g:, 'did_tex_path_update')
    return
  else
    let g:did_tex_path_update = 1
  endif

  let $PATH .= ':' . join([
        \ '/my/new/path/to/zathura',
        \ '/my/new/path/to/mupdf',
        \], ':')
endfunction
```

**Comment by teto** [comment]
Thanks for the answer. I perfectly get your point (simplicity is also why I chose vimtex over latexbox) and you suggested a few workarounds so I'll try to come up with a solution based on those. Thanks for your work.

**Comment by teto** [comment]
So I managed to workaround this specific issue by generating the .latexmkrc (see https://github.com/lervag/vimtex/issues/1073) `$bibtex="/nix/store/y6bn4fly4w71w7nhhnsmsb6qgfmrm7hl-texlive-combined-2017/bin/bibtex"` but on every startup I have
```
vimtex: bibtex is not executable!                                                                                                       
        bibtex completion is not available!
```
which is annoying. Any idea how to hide that ?
Maybe vimtex could instead move the warnings to neovim `:checkhealth` with advice on potential solving steps. I don't know if it would be compatible with the vim CheckHealth plugin though (but checkhealth is a killerfeature IMO).

**Comment by teto** [comment]
I guess I should have looked at the main.log too.
Also related to the linked issue. VimtexInfo could advise to run VimtexClean as the cache was a source of error in that issue (VimtexInfo might show the current values but not necessarily the ones used by latex).

**Comment by lervag** [comment]
The most simple way to hide that would be to somehow add bibtex to your `$PATH`. I'm not quite sure if it is possible to hide the warning in any other manner right now. But before I add a new option or update the code, I would be interested in understanding why one would ever want a LaTeX installation without bibtex. 

I agree `:checkhealth` could be relevant here. Do you now if the `CheckHealth` plugin in Vim is compatible with the neovim version? I am only interested if this would somehow work on both Vim and neovim. Note, an issue with `:checkhealth` is that it would only provide warnings when the user runs the command. This could potentially mean more issues raised because of lack of warnings.

I don't want to add any advise on running `VimtexClean`, as this is not really vimtex related but rather LaTeX related. I think users of vimtex must be expected to know their systems and basic(+) LaTeX.

**Comment by teto** [comment]
> The most simple way to hide that would be to somehow add bibtex to your $PATH. I'm not quite sure if it is possible to hide the warning in any other manner right now. But before I add a new option or update the code, I would be interested in understanding why one would ever want a LaTeX installation without bibtex.

It's not because it's not in your path that it does not exist :) One of the benefits of nixos is that you can run several versions of the same programs/libraries in parallel (out of the box virtualization). To achieve this, all paths are made explicit and not having binaries in the PATH helps in doing so (since programs error when they can't find a binary, and thus show where to set the absolute paths).
For instance some people don't install firefox system wide but just run `nix-shell -p firefox` and in parallel can run another version of firefox `nix-shell -p firefox-unstable` (achieved just using absolute paths, there is no qemu/docker overhead).
Adding a (fake) bibtex in path would defeat the purpose. Even on traditional system, if someone want to use its own version of a program, he might not want it to put into PATH because it's unstable.

I don't know if CheckHealth on vim behaves the same as on neovim but the author https://github.com/rhysd/vim-healthcheck has a long history of doing good work so I would be pretty optimistic about it.
I believe checkhealth is great to check for ambiguous/dubious problems. You can display a warning that is useful for a beginner and easily ignored by more expert people.

> I don't want to add any advise on running VimtexClean, as this is not really vimtex related but rather LaTeX related. I think users of vimtex must be expected to know their systems and basic(+) LaTeX.

I disagree on this. Users have no choice than knowing latex since using it is hard (multipass compilation, so many package incompatibilites, I dream of having namespaces) but most people would be happier knowing less about it (I know I do). There are many compilers/different versions of latex, because of some publisher you are forced to use packages you are not used to the complexity is high. Anything that helps is welcome in my book. Adding a notice about running VimtexClean prior to running VimtexInfo in VimtexInfo seems like a small price to pay. Typically if I had opened an issue with my VimtexInfo output, we could not have diagnosed my problem because it was not the configuration used at the moment by latex. At the same time it's true it might be an edge case and I tend to go the verbose way too often xD

**Comment by lervag** [comment]
Ok, now I feel we are talking about three different things in the same thread. I therefore propose the following:

1. We continue the discussion about bibtex here.
2. You open a new issue with the suggestion about CheckHealth (I agree this is nice, and it might help clean up warning messages).
3. You open a new issue on your suggestion for mentioning VimtexClean. To be clear: I do not quite understand exactly what you are suggesting here. So if you could be very clear about what you mean, that would help. Perhaps I misunderstood the first time.

----

So, back to bibtex. Do I understand correctly that what you need on nixos is a way to explicitly set every executable? Currently, vimtex assumes that executables are available from ones `$PATH`, and this works well on Linux (in general), OSX and Windows.

**Comment by teto** [comment]
I agree. As you suggested, I use .latexmkrc instead to set the paths for bibtex. I am fine with that. I just wish that the warning on nvim startup about "bibtex not in PATH" disappears. Such a warning could be moved to a checkhealth script.

**Comment by lervag** [comment]
So, if I understand correctly, these two things would satisfy your needs:

1. A way to silence (specific) warnings, and

2. A way to specify the bibtex executable.

If I am right, then I think both of these sound reasonable. I don't have that much time on my hands, but I will definately look into it when I get the time.

**Comment by teto** [comment]
I ended up writing a shell.nix to create a shell with bibtex/latexmk in PATH.

I put it here not for my future self and other nixos-ers
```
 { pkgs ? import <nixpkgs> {} }:
with pkgs;
let
  texliveEnv = pkgs.texlive.combine {
    inherit (pkgs.texlive) scheme-small cleveref latexmk bibtex algorithms cm-super
    csvsimple subfigure  glossaries natbib biblatex logreq xstring  subfiles 
    xfor mfirstuc datatool substr;
   };
in
stdenv.mkDerivation {
  name = "myGreatPaper";
  buildInputs = [ texliveEnv simpleBuildTool ];
}
```
I guess this is more in the spirit of nixos rather than installing latexmk systemwide as I used to do. It means I can work around 2/ (if you feel it's too much work or that it bloats vimtex, feel free to not do it).

1/ would be really nice. As soon as I open a file outside of of my nix-shell, then I get the 2 warnings => 2 keypresses. Also I don't get why I have to press Enter after running :VimtexCompile ? Is it due to my configuration ? I just typed the command so I know I launched it, why should I confirm with Enter xD
```
vimtex: Compiler started in continuous mode                                                                                                                                      
Appuyez sur ENTRÉE ou tapez une commande pour continuer
```

**Comment by lervag** [comment]
> I ended up writing a shell.nix to create a shell with bibtex/latexmk in PATH.

Ah, nice! I think this is neat, and if you don't mind I'll consider this a sufficient solution for my point number 2.

> 1/ would be really nice. As soon as I open a file outside of of my nix-shell, then I get the 2 warnings => 2 keypresses. Also I don't get why I have to press Enter after running :VimtexCompile ? Is it due to my configuration ? I just typed the command so I know I launched it, why should I confirm with Enter xD (EDIT it doesn't seem to always happen)

I don't have this problem with having to hit Enter. I think a relevant option is `laststatus`, see `:h 'laststatus'`. I've set it to 2 (except on neovim where this is the default).

But I'll look into this. I'll add an option to toggle the verbosity.

**Comment by lervag** [comment]
I've added two options:

```
*g:vimtex_log_ignore*
  A list of regexes to filter info, warning, and error messages. If a logged
  message matches any of the regexes in this list, the message will not be
  printed to screen.

  Note: All messages may still be viewed with |VimtexLog|.

  Default: []

*g:vimtex_log_verbose*
  Whether or not to print messages to screen. Should generally be on, but may
  be turned off e.g. for debugging or testing purposes.

  Default: 1
```

You should be able to use the `g:vimtex_log_ignore` to ignore specific warnings as desired.

**Comment by teto** [comment]
I just set g:vimtex_log_verbose to 0 for now and it seems to work thanks :D

**Comment by lervag** [comment]
No problem! Note that this will prevent any message from being displayed. It could be better to simply ignore the specific messages that you find annoying.
